### PR TITLE
Add Stream Support in File SDK

### PR DESCRIPTION
**Pull request by slreznit**
Created a forked IStream implementation for File SDK
All UT passed locally

### CODE REVIEW

File: sdk/file_sdk/Common/exceptions.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#ifndef FILE_SDK_EXCEPTIONS_H
+#define FILE_SDK_EXCEPTIONS_H
+
+#include <exception>
```
Comment (shenhavba) [inline sdk/file_sdk/Common/exceptions.h:4]: there is no namespace #Resolved

File: sdk/file_sdk/Common/file_format.h [hunk 2]
```diff
@@ -17,16 +17,16 @@ class FileFormat : public IFileFormat
 
   virtual void SetTags(const vector<Tag>& tags);
 
-  virtual void Commit(shared_ptr<IStream> file, string& newExtension) = 0;
+  virtual void Commit(std::shared_ptr<IStream> file, string& newExtension) = 0;
```
Comment (shenhavba) [inline sdk/file_sdk/Common/file_format.h]: >std:: [](start = 22, length = 5)

why did you add std:: if there is using std::shared_ptr; ? #Resolved

File: sdk/file_sdk/Common/std_stream_adapter.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+  auto fail = mInputStream->fail();
+  // Windows implementation of StringStream has a bug
+  // that for an empty stream fail bit is set after calling seekg(0).
+  // Workaround the bug by calling clear() which clears the failures.
+  if(fail){
+    mInputStream->clear();
+  }
+  mInputStream->read(reinterpret_cast<char *>(buffer), bufferLength);
```
Comment (shenhavba) [inline sdk/file_sdk/Common/std_stream_adapter.cpp:39]: >reinterpret_cast [](start = 21, length = 16)

better you use static_cast #WontFix

File: sdk/file_sdk/Common/std_stream_adapter.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    auto currentPosition =  mOutputStream->tellp();
+    mOutputStream->seekp(0, std::ios_base::end);
+    position =  static_cast<int>(mOutputStream->tellp());
+    mOutputStream->seekp(currentPosition);
+    mOutputStream->clear();
+  }
+  return static_cast<uint64_t>(position);
+}
+
+void StdStreamAdapter::Size(uint64_t) {}
```
Comment (shenhavba) [inline sdk/file_sdk/Common/std_stream_adapter.cpp]: why there is no implementation to this function? #WontFix

File: sdk/file_sdk/Common/std_stream_adapter.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+  static std::shared_ptr<IStream> Create(std::shared_ptr<std::ostream> stdOutputStream) {
+    return static_pointer_cast<IStream>(make_shared<StdStreamAdapter>(stdOutputStream));
+  }
+
+  static std::shared_ptr<IStream> Create(std::shared_ptr<std::iostream> stdStream) {
+    return static_pointer_cast<IStream>(make_shared<StdStreamAdapter>(stdStream));
+  }
+
+//  static StdStreamAdapter Create(std::shared_ptr<std::stringstream> stdStringStream) {
```
Comment (shenhavba) [inline sdk/file_sdk/Common/std_stream_adapter.h]: you left the comment on purpose ? #Resolved

File: sdk/file_sdk/CompoundFileFormat/compound_file_format.cpp [hunk 6]
```diff
@@ -3,7 +3,7 @@
 namespace mip {
 namespace file {
 
-CompoundFileFormat::CompoundFileFormat(shared_ptr<IStream> file, const std::string& extension)
+CompoundFileFormat::CompoundFileFormat(std::shared_ptr<IStream> file, const std::string& extension)
```
Comment (shenhavba) [inline sdk/file_sdk/CompoundFileFormat/compound_file_format.cpp]: >std:: [](start = 39, length = 5)

why do you add it? #Resolved

**Review by shenhavba — CHANGES_REQUESTED** [review]
:clock1:

File: sdk/file_sdk/Common/std_stream_adapter.cpp (hunk 4 above)
Comment (slreznit) [reply sdk/file_sdk/Common/std_stream_adapter.cpp]: There seems to be no way to truncate/setSize of the underlying std::ostream.
Let's see if we nee dthis implementation and take it separately.

---
In reply to: [138536418](https://github.com/AzureAD/rms-sdk-for-cpp/pull/151#discussion_r138536418) [](ancestors = 138536418)

File: sdk/file_sdk/Common/std_stream_adapter.cpp (hunk 3 above)
Comment (slreznit) [reply sdk/file_sdk/Common/std_stream_adapter.cpp:39]: cannot use static_cast from unsigned to signed char

---
In reply to: [138536249](https://github.com/AzureAD/rms-sdk-for-cpp/pull/151#discussion_r138536249) [](ancestors = 138536249)

Comment (shenhavba) [reply sdk/file_sdk/Common/std_stream_adapter.cpp:39]: why is it signed char? shouldn't it be unsigned char? #Pending

Comment (slreznit) [reply sdk/file_sdk/Common/std_stream_adapter.cpp:39]: this is what read excepts

---
In reply to: [138578417](https://github.com/AzureAD/rms-sdk-for-cpp/pull/151#discussion_r138578417) [](ancestors = 138578417)

**Review by shenhavba — APPROVED** [review]
:shipit:
